src/Bicep.Core/TypeSystem/Providers/ResourceTypesProviderDescriptor.cs

```diff
@@ -1,33 +1,51 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Registry;
+using Bicep.Core.Semantics.Namespaces;
+
 namespace Bicep.Core.TypeSystem.Providers;
 
 public record ResourceTypesProviderDescriptor
 {
-    public ResourceTypesProviderDescriptor(
-        string name,
-        string version,
-        bool isImplicitImport,
-        string? alias = null,
-        string? artifactRef = null,
-        Uri? typesBaseUri = null)
+    public const string LegacyVersionPlaceholder = "1.0.0";
+
+    private ResourceTypesProviderDescriptor(
+        string NamespaceIdentifier,
+        string Version,
+        bool IsBuiltIn,
+        Uri ParentModuleUri,
+        string? Alias = null,
+        ArtifactReference? ArtifactReference = null,
+        Uri? TypesTgzUri = null)
     {
-
-        Name = name;
-        Alias = alias ?? name;
-        ArtifactReference = artifactRef ?? (isImplicitImport ? name : $"{name}@{version}");
-        Version = version;
-        TypesBaseUri = typesBaseUri;
+        this.Name = NamespaceIdentifier;
+        this.ParentModuleUri = ParentModuleUri;
+        this.Version = Version;
+        this.IsBuiltIn = IsBuiltIn;
+        this.Alias = Alias ?? NamespaceIdentifier;
+        this.ArtifactReference = ArtifactReference;
+        this.TypesTgzUri = TypesTgzUri;
     }
 
-    public string Name { get; }
+    public readonly ArtifactReference? ArtifactReference;
+    public readonly bool IsBuiltIn;
+    public readonly string Alias;
+    public readonly string Name;
+    public readonly Uri ParentModuleUri;
+    public readonly string Version;
+    public readonly Uri? TypesTgzUri;
```

Comment:
Is there a specific reason we opt for public readonly fields over properties? Generally, exposing public readonly fields isn't a common practice in C#.